### ISSUE TITLE
Ability to pass **kwargs to on_completion and on_failure hooks

### ISSUE THREAD

**Issue by prefectcboyd** [PrefectHQ/prefect#8884]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

New features in version 2.8.3 allow for hooks on task and flow completion, and task / flow failure.

@jakekaplan shared the link below to the source for the task implementation:
https://github.com/PrefectHQ/prefect/blob/main/src/prefect/engine.py#L1998-L2028

This unfortunately does not allow accessing any of the task / flow context parameters, or allow passing of any variables outside of the defined `task`, `task_run`, `state`. 

A contrived example of the relevance here - 
I run a flow, and upon completion, or failure, for change management purposes I would like to update a Jira or ServiceNow ticket. In order to do so, I need the flow name, a callback url, a secret, and a piece of flow identifying information that was an input parameter to start the flow.

I call this function like:
`def update_api(callback_url, step_id, execution_id, secret)`.

This would be the same call for success or failure, with the only difference being the result / state update. 

Using `on_completion` or `on_failure` however, do not allow me to pass these variables, or access them. E.g.
`@flow(on_completion=[update_api])` - I have all the variables accessible in flow_context , and within the flow, but I don't have a way to pass them into the hook.

### Describe the proposed behavior

Ideally, a modification to allow **kwargs to be passed into the hooks which can be unpacked with whatever inputs are necessary for the hook.

### Example Use

Using the given example above (with callback_url etc.)

```
def update_api(callback_url, step_id, execution_id, secret):
...
...

@flow(on_completion=[update_api(callback_url, step_id, execution_id, secret)]
```

Alternatively, a dictionary example:
```
def update_api(params):
...
...

@flow(on_completion=[update_api(params)]
params['callback_url'] = 'https://my-callback.com'
params['step_id'] = 'aab1-eeac'
params['execution_id'] = '867-5309'
params['secret'] = 'abc123'
```

**Comment by serinamarie** [comment]
@billpalombi I'm wondering if we should support something like this?

**Comment by billpalombi** [comment]
I've seen two requests for this enhancement. I like it. As long as it doesn't add a lot of complexity, I fully support it.

**Comment by billpalombi** [comment]
[Related discourse thread](https://discourse.prefect.io/t/passing-and-retrieving-data-in-the-on-failure-hook/2541)

**Comment by cicdw** [comment]
Just saw the open PR for this issue and I don't think I'm clear on what we have to solve here?  

For dynamic inputs, users can use `functools.partial`, and for accessing runtime information users can import `prefect.runtime` within their hook - that should cover both of the use cases in the discourse topic if I'm reading correctly.

**Comment by prefectcboyd** [comment]
At the time of drafting this , I don't believe prefect.runtime was available as it was released in 2.9.0.
I do think that does solve the issue here in being able to access runtime parameters in hook, so that should work.
Thanks!
